File: backend/src/calendar_utils.py
from google.auth.transport.requests import Request
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError
import datetime
import os
from dotenv import load_dotenv  # Load environment variables

# Load environment variables from .env
load_dotenv()

# Set paths for persistent storage
TOKEN_PATH = r"config/token.json"
CREDENTIALS_PATH = r"config/credentials.json"

SCOPES = ["https://www.googleapis.com/auth/calendar.readonly"]

def get_calendar_events():
    """Fetches events up to 90 days from Google Calendar and ensures token persistence."""
    creds = None

    # Debug: Check if GOOGLE_CALENDAR_ID is loaded
    calendar_id = os.getenv("GOOGLE_CALENDAR_ID")
    if not calendar_id:
        raise ValueError("❌ GOOGLE_CALENDAR_ID is not set! Check your .env file.")

    print(f"📢 Using Calendar ID: {calendar_id}")

    # Load token.json if it exists
    if os.path.exists(TOKEN_PATH):
        print("✅ Found token.json, loading credentials.")
        creds = Credentials.from_authorized_user_file(TOKEN_PATH, SCOPES)

    # Refresh or generate a new token if needed
    if not creds or not creds.valid:
        if creds and creds.expired and creds.refresh_token:
            print("🔄 Refreshing expired token...")
            creds.refresh(Request())
        else:
            print("🚨 No valid credentials found, running OAuth flow.")
            flow = InstalledAppFlow.from_client_secrets_file(CREDENTIALS_PATH, SCOPES)
            creds = flow.run_local_server(port=0)

        # Store the refreshed token
        with open(TOKEN_PATH, "w") as token:
            token.write(creds.to_json())

    try:
        print("🔗 Connecting to Google Calendar API...")
        service = build("calendar", "v3", credentials=creds)
        
        now = datetime.datetime.utcnow().isoformat() + "Z"
        time_max = (datetime.datetime.utcnow() + datetime.timedelta(days=90)).isoformat() + "Z"
        
        events_result = service.events().list(
            calendarId=calendar_id,  # Now we know this is set correctly
            timeMin=now,
            timeMax=time_max,
            singleEvents=True,
            orderBy="startTime",
        ).execute()

        events = events_result.get("items", [])
        print(f"📆 Retrieved {len(events)} events from Google Calendar.")

        return events

    except HttpError as error:
        print(f"❌ Google API Error: {error}")
        return []

File: lib/calendar_utils.js
import { google } from 'googleapis';
import fs from 'fs/promises';
import path from 'path';

const TOKEN_PATH = path.join(process.cwd(), 'config', 'token.json');
const CREDENTIALS_PATH = path.join(process.cwd(), 'config', 'credentials.json');
const SCOPES = ['https://www.googleapis.com/auth/calendar.readonly'];

async function loadSavedCredentialsIfExist() {
  try {
    const content = await fs.readFile(TOKEN_PATH, 'utf8');
    const credentials = JSON.parse(content);
    return google.auth.fromJSON(credentials);
  } catch (err) {
    return null;
  }
}

async function saveCredentials(client) {
  const content = await fs.readFile(CREDENTIALS_PATH, 'utf8');
  const keys = JSON.parse(content);
  const key = keys.installed || keys.web;
  
  const payload = JSON.stringify({
    type: 'authorized_user',
    client_id: key.client_id,
    client_secret: key.client_secret,
    refresh_token: client.credentials.refresh_token,
    access_token: client.credentials.access_token,
    token_type: client.credentials.token_type,
    expiry_date: client.credentials.expiry_date,
  });
  
  await fs.writeFile(TOKEN_PATH, payload);
}

async function authorize() {
  let client = await loadSavedCredentialsIfExist();
  
  if (client) {
    console.log('✅ Found token.json, loading credentials.');
    
    if (client.credentials.expiry_date && client.credentials.expiry_date < Date.now()) {
      console.log('🔄 Refreshing expired token...');
      try {
        await client.refreshAccessToken();
        await saveCredentials(client);
      } catch (error) {
        console.error('❌ Error refreshing token:', error);
        return null;
      }
    }
    return client;
  }
  
  console.log('🚨 No valid credentials found.');
  return null;
}

export async function getCalendarEvents() {
  const calendarId = process.env.GOOGLE_CALENDAR_ID;
  
  if (!calendarId) {
    throw new Error('❌ GOOGLE_CALENDAR_ID is not set!');
  }
  
  console.log(`📢 Using Calendar ID: ${calendarId}`);
  
  const auth = await authorize();
  
  if (!auth) {
    throw new Error('Failed to authorize with Google Calendar API');
  }
  
  try {
    console.log('🔗 Connecting to Google Calendar API...');
    const calendar = google.calendar({ version: 'v3', auth });
    
    const now = new Date();
    const timeMax = new Date();
    timeMax.setDate(timeMax.getDate() + 90);
    
    const response = await calendar.events.list({
      calendarId: calendarId,
      timeMin: now.toISOString(),
      timeMax: timeMax.toISOString(),
      singleEvents: true,
      orderBy: 'startTime',
    });
    
    const events = response.data.items || [];
    console.log(`📆 Retrieved ${events.length} events from Google Calendar.`);
    
    return events;
  } catch (error) {
    console.error('❌ Google API Error:', error);
    throw error;
  }
}